halo2-base/src/poseidon/hasher/mod.rs

```diff
@@ -16,15 +22,142 @@ pub mod spec;
 /// Module for poseidon states.
 pub mod state;
 
-/// Poseidon hasher. This is stateful.
+/// Stateless Poseidon hasher.
 pub struct PoseidonHasher<F: ScalarField, const T: usize, const RATE: usize> {
+    spec: OptimizedPoseidonSpec<F, T, RATE>,
+    consts: OnceCell<PoseidonHasherConsts<F, T, RATE>>,
+}
+#[derive(Getters)]
+struct PoseidonHasherConsts<F: ScalarField, const T: usize, const RATE: usize> {
+    #[getset(get = "pub")]
+    init_state: PoseidonState<F, T, RATE>,
+    // hash of an empty input("").
+    #[getset(get = "pub")]
+    empty_hash: AssignedValue<F>,
+}
+
+impl<F: ScalarField, const T: usize, const RATE: usize> PoseidonHasherConsts<F, T, RATE> {
+    pub fn new(
+        ctx: &mut Context<F>,
+        gate: &impl GateInstructions<F>,
+        spec: &OptimizedPoseidonSpec<F, T, RATE>,
+    ) -> Self {
+        let init_state = PoseidonState::default(ctx);
+        let mut state = init_state.clone();
+        let empty_hash = fix_len_array_squeeze(ctx, gate, &vec![], &mut state, spec);
```

Comment:
i guess this could in principle be precomputed using `Constant` instead of `Existing`. if it's a one-time cost then it doesn't matter

Reply:
I think we only call this once so it should be fine.